samples/sample_common/src/sample_utils.cpp

```diff
@@ -1213,6 +1213,109 @@ mfxStatus CSmplYUVWriter::WriteNextFrameI420(mfxFrameSurface1 *pSurface)
     return MFX_ERR_NONE;
 }
 
+QPFileReader::QPFileReader()
+{
+    m_nframes = 0;
+    m_cur_frame_num = 0;
+}
+
+QPFileReaderStatus QPFileReader::Read(const msdk_string &strFileName, mfxU32 codecid)
+{
+    msdk_fstream ifs;
+    ifs.open(strFileName, msdk_fstream::in);
+
+    if (ifs.is_open())
+    {
+        QPFileReaderFrameInfo frame_info;
+        msdk_string frame_type{};
+
+        if (!(ifs >> m_nframes))
+            return QPFILEREADER_WRONG_FILE;
+
+        m_frame_info.reserve(m_nframes);
+
+        if (codecid == MFX_CODEC_AVC)
+        {
+            while (ifs >> frame_info.display_order_num && ifs >> frame_info.QP && ifs >> frame_type)
+            {
+                if (frame_type == MSDK_STRING("IDR_REF"))
+                    frame_info.frame_type = MFX_FRAMETYPE_I | MFX_FRAMETYPE_IDR | MFX_FRAMETYPE_REF;
+                else if (frame_type == MSDK_STRING("P_REF"))
+                    frame_info.frame_type = MFX_FRAMETYPE_P | MFX_FRAMETYPE_REF;
+                else if (frame_type == MSDK_STRING("B"))
+                    frame_info.frame_type = MFX_FRAMETYPE_B;
+                else if (frame_type == MSDK_STRING("B_REF"))
+                    frame_info.frame_type = MFX_FRAMETYPE_B | MFX_FRAMETYPE_REF;
+                else
+                    return QPFILEREADER_WRONG_FILE;
+                m_frame_info.push_back(frame_info);
+            }
+        }
+        else if (codecid == MFX_CODEC_HEVC)
+        {
+            while (ifs >> frame_info.display_order_num && ifs >> frame_info.QP && ifs >> frame_type)
+            {
+                if (frame_type == MSDK_STRING("IDR_REF"))
+                    frame_info.frame_type = MFX_FRAMETYPE_IDR | MFX_FRAMETYPE_I | MFX_FRAMETYPE_REF;
+                else if (frame_type == MSDK_STRING("I_REF"))
+                    frame_info.frame_type = MFX_FRAMETYPE_I | MFX_FRAMETYPE_REF;
+                else if (frame_type == MSDK_STRING("P_REF"))
+                    frame_info.frame_type = MFX_FRAMETYPE_P | MFX_FRAMETYPE_REF;
+                else if (frame_type == MSDK_STRING("P"))
+                    frame_info.frame_type = MFX_FRAMETYPE_P;
+                else if (frame_type == MSDK_STRING("B_REF"))
+                    frame_info.frame_type = MFX_FRAMETYPE_B | MFX_FRAMETYPE_REF;
+                else if (frame_type == MSDK_STRING("B"))
+                    frame_info.frame_type = MFX_FRAMETYPE_B;
+                else
+                    return QPFILEREADER_WRONG_FILE;
+                m_frame_info.push_back(frame_info);
+            }
+        }
+        else return QPFILEREADER_CODEC_UNSUPPORTED;
+
+        if (m_frame_info.size() < m_nframes)
+            return QPFILEREADER_WRONG_FILE;
+
+        for (const QPFileReaderFrameInfo &finfo : m_frame_info)
+            if (finfo.QP < 1 || finfo.QP > 51)
```

Comment:
what about QP == 0 handling? in HEVC standard it is valid value... We had some problems with handling it internally. Anyway, need to clarify this